### PR TITLE
[iOS] Enable sliding tabs on TabbedPage with page-level configurability

### PR DESCRIPTION
**Pull request by adrianknight89**
### Description of Change

Could be marked as enhancement. This PR will let you make iOS `TabbedPage` richer by enabling pager-like behavior. Unlike Android, iOS `UITabBarController` doesn't have support for sliding tabs. `TabSwipe` is a dictionary of `SwipeConfig` objects that let you configure sliding behavior at a page level. You can enable/disable sliding on a particular page, set slide animation duration, or set required velocity to slide pages without them reaching the middle of the screen.

Here's a demo: https://1drv.ms/v/s!AjlbPgOcTyP2bPDoCzlPPYkfw_s
### API Changes

Added:
- public static Dictionary<int, TabSwipe> GetTabSwipes(BindableObject element)
- public static void SetTabSwipes(BindableObject element, Dictionary<int, TabSwipe> value)
- public static Dictionary<int, TabSwipe> TabSwipes(this IPlatformElementConfiguration<iOS, FormsElement> config)
- public static IPlatformElementConfiguration<iOS, FormsElement> SetTabSwipes(this IPlatformElementConfiguration<iOS, FormsElement> config, Dictionary<int, TabSwipe> value)
- public static IPlatformElementConfiguration<iOS, FormsElement> EnableTabSwipes(this IPlatformElementConfiguration<iOS, FormsElement> config)
- public static IPlatformElementConfiguration<iOS, FormsElement> EnableTabSwipes(this IPlatformElementConfiguration<iOS, FormsElement> config)
### PR Checklist
- [ ] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense

### CODE REVIEW

**Comment by oddbear** [comment]
Nice, but why those:
```csharp
while (!_someBoolean)
{
    await Task.Delay(5);
}
```
with no code after, inside <code>async void</code>?

**Comment by adrianknight89** [comment]
@oddbear `UIView.Animate` returns immediately before finishing (i.e. the completion callback is raised when the animation is done but execution does not wait.) That code is basically a hack. I think it should be similar to [Platform](https://github.com/xamarin/Xamarin.Forms/blob/53e1d99a751e7a9cb8ae3f67ac8f3630b2aea398/Xamarin.Forms.Platform.iOS/Platform.cs#L461), which I also worked on in my previous PR.

If I do not await the run-to-completion or snap-into-position animations, then it potentially messes up slides if you create pan gestures like a child while those animations are running. 

Is there another animation method we could use?

File: Xamarin.Forms.Platform.iOS/Renderers/TabbedRenderer.cs [hunk 1]
```diff
@@ -387,6 +401,297 @@ void UpdateCurrentPage()
...
+				ShouldSelectViewController = null;
+
+				if (_uiPanGestureRecognizer == null)
+					return;
+
+				if(View.GestureRecognizers.Contains(_uiPanGestureRecognizer))
+					View.GestureRecognizers.Remove(_uiPanGestureRecognizer);
+
+				_uiPanGestureRecognizer.Dispose();
+				_uiPanGestureRecognizer = null;
```
Comment (adrianknight89) [inline Xamarin.Forms.Platform.iOS/Renderers/TabbedRenderer.cs]: Removing and disposing `_uiPanGestureRecognizer` doesn't seem to be removing the target/handler (i.e. the handler gets called still when the object is disposed and set to null) Does anyone know why? GC should be relatively quick for cleanup. Not sure if this is a bug.

**Comment by adrianknight89** [comment]
Ideally, the animations should be abortable with a pan gesture so the gesture controls slides. This will require a future update.

File: Xamarin.Forms.Platform.iOS/Renderers/TabbedRenderer.cs (hunk 1 above)
Comment (Emasoft) [inline Xamarin.Forms.Platform.iOS/Renderers/TabbedRenderer.cs]: Take a look at this code: https://github.com/xmartlabs/XLPagerTabStrip

**Comment by rmarinho** [comment]
needs rebase

**Comment by rmarinho** [comment]
@adrianknight89 Can you rebase please! Thanks

**Comment by adrianknight89** [comment]
Done.

**Comment by adrianknight89** [comment]
After the direction XF has taken, I think it's probably better to have this as a 3rd party component or people should be encouraged to use `CarouselView` when it's out.

**Comment by brittonbeckham** [comment]
I vote to reopen this as the new direction is to adapt things like this

**Pull request closed without merging**